portal/src/util/error.ts

```diff
@@ -0,0 +1,175 @@
+    return null;
+  }
+  return {
+    kind: cause.kind,
+    missingField: cause.details.missing,
+    location: cause.location,
+  };
+}
+
+function handleUpdateAppConfigError(error: GraphQLError): Violation[] {
+  if (!isGraphQLErrorExtensions(error.extensions)) {
+    return [];
+  }
+  const causes = error.extensions.info.causes ?? [];
+
+  return causes.map(extractViolationFromErrorCause).filter(nonNullable);
+}
+
+export function parseError(error: unknown): Violation[] {
+  if (error instanceof ApolloError) {
+    const violations: Violation[] = [];
+    for (const graphQLError of error.graphQLErrors) {
+      const errorViolations = handleUpdateAppConfigError(graphQLError);
+      for (const violation of errorViolations) {
+        violations.push(violation);
+      }
+    }
+    return violations;
+  }
+
+  // unrecognized error
+  return [];
+}
+
+// default seclectors
+export function makeMissingFieldSelector(
+  locationPrefix: string,
+  fieldName: string
+): ViolationSelector {
+  return function (violation: Violation) {
+    /* uncomment when there is more than one violation kind
+    if (violation.kind !== "required") {
+      return false
+    }
+    */
+    if (!violation.location.startsWith(locationPrefix)) {
+      return false;
+    }
+    return violation.missingField.includes(fieldName);
+  };
+}
+
+export function combineSelector(
+  selectorList: ViolationSelector[]
+): ViolationSelector {
+  return function (violation: Violation) {
+    for (const selector of selectorList) {
+      if (selector(violation)) {
+        return true;
+      }
+    }
+    return false;
+  };
+}
+
+export function violationSelector<Key extends string>(
+  violations: Violation[],
+  violationSelectors: ViolationSelectors<Key>
+): Record<Key, Violation[]> {
+  const violationMap = Object.entries(violationSelectors).reduce<
+    Partial<Record<Key, Violation[]>>
+  >((violationMap, [key, selector]) => {
+    violationMap[key as Key] = violations.filter(selector as ViolationSelector);
+    return violationMap;
+  }, {});
+  return violationMap as Record<Key, Violation[]>;
+}
+
+function violationFormatter(
+  fieldNameId: string,
+  violation: Violation,
+  renderToString: (messageId: string, values?: Values) => string
+): string | undefined {
+  switch (violation.kind) {
+    case "required":
+      return renderToString("required-field-missing", {
+        fieldName: renderToString(fieldNameId),
+      });
+  }
+  return undefined;
+}
+
+export function errorFormatter(
+  fieldNameId: string,
+  violations: Violation[],
+  renderToString: (messageId: string, values?: Values) => string,
+  formatErrorMessageList: (
+    errorMessages: string[]
+  ) => string | undefined = defaultFormatErrorMessageList
+): string | undefined {
+  return formatErrorMessageList(
+    violations
+      .map((violation) =>
+        violationFormatter(fieldNameId, violation, renderToString)
+      )
+      .filter(nonNullable)
+  );
+}
```

Comment:
Can you write tests to demonstrate the usage? With tests it is more much easier to see how everything interacts. The tests do not need to rely on React.